src/BagUtils.php

```diff
@@ -353,4 +353,66 @@ public static function checkedFwrite($fp, $content)
             throw new FilesystemException("Error writing to file");
         }
     }
+
+    /**
+     * Decode a file path according to the special rules of the spec.
+     *
+     * RFC 8943 - sections 2.1.3 & 2.2.3
+     * If _filename_ includes an LF, a CR, a CRLF, or a percent sign (%), those characters (and only those) MUST be
+     * percent-encoded as described in [RFC3986].
+     *
+     * @param string $line
+     *   The original filepath from the manifest file.
+     * @return string
+     *   The filepath with the special characters decoded.
+     */
+    public static function decodeFilepath(string $line): string
+    {
+        // Strip newlines from the right.
+        $decoded = rtrim($line, "\r\n");
+        return str_replace(
+            ["%0A", "%0D", "%25"],
+            ["\n", "\r", "%"],
+            $decoded
+        );
+    }
+
+    /**
+     * Encode a file path according to the special rules of the spec.
+     *
+     * RFC 8943 - sections 2.1.3 & 2.2.3
+     * If _filename_ includes an LF, a CR, a CRLF, or a percent sign (%), those characters (and only those) MUST be
+     * percent-encoded as described in [RFC3986].
+     *
+     * @param string $line
+     *   The original file path.
+     * @return string
+     *   The file path with the special manifest characters encoded.
+     */
+    public static function encodeFilepath(string $line): string
+    {
+        // Strip newlines from the right.
+        $encoded = rtrim($line, "\r\n");
+        return str_replace(
+            ["%", "\n", "\r"],
+            ["%25", "%0A", "%0D"],
+            $encoded
+        );
+    }
+
+    /**
+     * Check for unencoded newlines, carriage returns or % symbols in a file path.
+     *
+     * @param string $filepath
+     *   The file path to check
+     * @return bool
+     *   True if there are un-encoded characters
+     * @see \whikloj\BagItTools\BagUtils::encodeFilepath()
+     * @see \whikloj\BagItTools\BagUtils::decodeFilepath()
+     */
+    public static function checkUnencodedFilepath(string $filepath): bool
+    {
+        return (strpos($filepath, "\n") > -1 || strpos($filepath, "\r") > -1 ||
```

Comment:
It's unclear to me what `fgets` considers a newline, but if you're reading the file line by line as defined by the spec where CR, LF, or CRLF are line endings, then it shouldn't be possible for the filepath to contain a CR or LF at this point.

The encoding check is technically correct. However, it does make it so the code would reject bags created by libraries that haven't be updated to encode paths correctly. I don't know if you're interested in accepting those bags or not though.

Reply:
Yeah I struggled with this, mostly because I don't use PHP on Windows and so I am biased towards LF as newline characters. Looking at it now, I'm not handling CR line endings properly. I'll make a new ticket for that.

Reply:
#36